src/mtr.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+  Adplug - Replayer for many OPL2/OPL3 audio file formats.
+  Copyright (C) 1999 - 2003 Simon Peter, <dn.tlp@gmx.net>, et al.
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+  mtr.cpp - MTR loader by Dmitry Smagin <dmitry.s.smagin@gmail.com>
+*/
+
+#include <stdio.h>
+#include <cstring>
+#include "mtr.h"
+#include "debug.h"
+
+/* -------- Public Methods -------------------------------- */
+
+CPlayer *CmtrLoader::factory(Copl *newopl) {
+    return new CmtrLoader(newopl);
+}
+
+bool CmtrLoader::load(const std::string &filename, const CFileProvider &fp) {
+    binistream *f = fp.open(filename);
+    if (!f)
+        return false;
+
+    unsigned int i, j, k, t = 0;
+    ninstruments = 0;
+
+    char header[51] = {0};
+    char mtitle[21] = {0};
+    int nvoices, ndigvoices, npatterns, orderlen,
+        timervalue = 0x428F, restart, flength;
+
+    f->readString(header, 50);
+
+    if (!strncmp(header, "MTRAC ", 6)) {
+        version = 1;
+        sscanf(header + 26, "%02x %02x %02x %02x %08x",
+            &nvoices,
+            &npatterns,
+            &orderlen,
+            &restart,
+            &flength
+        );
+        strncpy(mtitle, header + 6, 20);
+        timervalue = f->readInt(2); // usually 428F
+        f->ignore(1); // 0=SPK 1=ADL 2=SBP
+    } else if (!strncmp(header, "MTRACK NC", 9)) {
+        version = 2;
+        sscanf(header + 10, "%02x %02x %02x %02x %02x %02x %04x %08x",
```

Comment:
Need to error-out if sscan() returns another value than 8

Reply:
Okay, some error handling for sscanf won't hurt if the file is severely broken.